### PR TITLE
[HLS-381] New variant splitting transformer with behavior similar to vt decompose 

### PR DESCRIPTION
**Pull request by kianfar77**
## What changes are proposed in this pull request?
Glow's current multi-allelic variant splitting transformer (part of normalize_variants transformer) uses ``GATKVariantContextUtils.splitVariantContextToBiallelics`` function for splitting whose behavior is similar to GATK's ``LeftAlignAndTrimVariants --splitMultiallelics`` and has limitations in conversion of INFO and genotype-related fields. This PR implements a new variant splitter which behaves like ``vt decompose``. In addition the splitting happens directly in Spark SQL eliminating the transformation to VariantContext and back, which is expected to significantly improve performance. 

## How is this patch tested?
- [x] Unit tests
- [ ] Integration tests
- [x] Manual tests

### CODE REVIEW

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 1]
```diff
@@ -76,92 +76,106 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+            case ArrayType(_, _) => true
+            case _ => false
+          }))
+      .map(field =>
+        (f: DataFrame) => {
+          f.withColumn(
+            field.name,
+            when(
+              size(col(field.name)) === size(col(alternateAllelesField.name)),
+              expr(s"${field.name}[altAllelePos]")))
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: I don't think this is necessarily correct? If the number is `A` then it makes sense, but if the number is `R` then there's an entry for the reference allele as well.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 2]
```diff
@@ -67,92 +69,331 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+    // and posexplode alternateAlleles column
+    val dfAfterAltAlleleSplit = variantDf
+      .withColumn(
+        splitFromMultiAllelicField.name,
+        when(size(col(alternateAllelesField.name)) > 1, lit(true)).otherwise(lit(false))
+      )
+      .withColumn(
+        infoFieldPrefix + oldMultiallelicFieldName,
+        when(
+          expr(s"${splitFromMultiAllelicField.name} == true"),
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: nit: col(splitFromMultiAllelicField.name)

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 3]
```diff
@@ -67,92 +69,331 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+          concat(
+            col(s"${contigNameField.name}"),
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: col(contigNameField.name)

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 4]
```diff
@@ -67,92 +69,331 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+
+    // Update splitFromMultiAllelic column, add INFO_OLD_MULTIALLELIC column (see vt decompose)
+    // and posexplode alternateAlleles column
+    val dfAfterAltAlleleSplit = variantDf
+      .withColumn(
+        splitFromMultiAllelicField.name,
+        when(size(col(alternateAllelesField.name)) > 1, lit(true)).otherwise(lit(false))
+      )
+      .withColumn(
+        infoFieldPrefix + oldMultiallelicFieldName,
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Is this a standard field?

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 5]
```diff
@@ -67,92 +69,331 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+  private def splitGenotypeFields(variantDf: DataFrame): DataFrame = {
+
+    // get genotypes schema
+    val gSchema = variantDf
+      .schema
+      .fields
+      .find(_.name == genotypesFieldName)
+      .get
+      .dataType
+      .asInstanceOf[ArrayType]
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Throw an error if it's not an `ArrayType`

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 6]
```diff
@@ -67,92 +69,331 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+      .elementType
+      .asInstanceOf[StructType]
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Throw an error if it's not a `StructType`

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 4 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: This is a field that vt decompose adds to keep the multiallelic information of the split alleles. In discussion with @habeggel, it was mentioned that it is useful to add it. I intend to create an option for the user to be able to disable creation of this column.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 3 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: sorry, remnants of modified larger expressions!

File: build.sbt [hunk 7]
```diff
@@ -27,7 +27,7 @@ concurrentRestrictions in Global := Seq(
   Tags.limit(Tags.ForkedTestGroup, testConcurrency)
 )
 
-def groupByHash(tests: Seq[TestDefinition]) = {
+def groupByHash(tests: Seq[TestDefinition]): Seq[Tests.Group] = {
```
Comment (henrydavidge) [inline build.sbt]: Why this change?

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 8]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
    */
-  private def splitVC(vc: VariantContext): Stream[VariantContext] = {
-    GATKVariantContextUtils
-      .splitVariantContextToBiallelics(
-        vc,
-        false,
-        GenotypeAssignmentMethod.BEST_MATCH_TO_ORIGINAL,
-        false
+  @VisibleForTesting
+  private[normalizevariants] def splitVariants(variantDf: DataFrame): DataFrame = {
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Could we move the splitter into a separate file? It's pretty separate, and this file is getting long.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 9]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+                  f.name,
+                  when(
+                    col(splitFromMultiAllelicField.name),
+                    expr(
+                      s"transform(${f.name}, " +
+                      s"c -> transform(c, x -> if(x == 0, x, if(x == $splitAlleleIdxFieldName + 1, 1, -1))))"
+                    )
+                  ).otherwise(col(f.name))
+                )
+                .withColumn("contigName", expr("dummyUdf(contigName)"))
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Do we still need this `dummyUdf`?

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 10]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+
+    // get genotypes schema
+    val genotypesFieldDataType = variantDf
+      .schema
+      .fields
+      .find(_.name == genotypesFieldName)
+      .get
+      .dataType
+
+    if (!genotypesFieldDataType.isInstanceOf[ArrayType]) {
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: I think that @karenfeng added a utility function to validate and get the genotype schema.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 11]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+        when(size(col(alternateAllelesField.name)) > 1, lit(true)).otherwise(lit(false))
+      )
+      .withColumn(
+        infoFieldPrefix + oldMultiallelicFieldName,
+        when(
+          col(splitFromMultiAllelicField.name),
+          concat(
+            col(contigNameField.name),
+            lit(":"),
+            expr(s"${startField.name} + 1"),
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: nit: `col(startField.name) + 1`

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 12]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
-        vc,
-        false,
-        GenotypeAssignmentMethod.BEST_MATCH_TO_ORIGINAL,
-        false
+  @VisibleForTesting
+  private[normalizevariants] def splitVariants(variantDf: DataFrame): DataFrame = {
+
+    if (variantDf.schema.fieldNames.contains("attributes")) {
+      // TODO: Unflattened INFO field splitting
+      logger.warn(
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: In Databricks, it could be nice to provide a Spark advice tip for this case.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 13]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+      .withColumn(
+        infoFieldPrefix + oldMultiallelicFieldName,
+        when(
+          col(splitFromMultiAllelicField.name),
+          concat(
+            col(contigNameField.name),
+            lit(":"),
+            expr(s"${startField.name} + 1"),
+            lit(":"),
+            concat_ws("/", col(s"${refAlleleField.name}"), col(s"${alternateAllelesField.name}"))
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: No need to use string interpolation here. Just `col(refAlleleField.name)`.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 14]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+
+    val genotypesFieldElementType = genotypesFieldDataType.asInstanceOf[ArrayType].elementType
+
+    if (!genotypesFieldElementType.isInstanceOf[StructType]) {
+      throw new ClassCastException("Genotypes field element is not of StructType.")
+    }
+
+    val gSchema = genotypesFieldElementType.asInstanceOf[StructType]
+
+    val dummyUdf = udf { x: String =>
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: I hope that we don't need this anymore, but if we do we need a comment.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 10 above)
Comment (karenfeng) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: You can find this in `InternalRowToVariantContextConverter.getGenotypeSchema`

File: build.sbt (hunk 7 above)
Comment (kianfar77) [reply build.sbt]: Not a big deal but without the explicit type intelliJ always shows a formatting error. This is just to get rid of that eyesore!

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 8 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Done

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 12 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Will do.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 10 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Replaced.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 14 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: My bad. Forgot to delete before push.

File: core/src/test/scala/io/projectglow/transformers/normalizevariants/NormalizeVariantsTransformerSuite.scala [hunk 15]
```diff
@@ -73,20 +74,26 @@ class NormalizeVariantsTransformerSuite extends GlowBaseTest with GlowLogging {
...
+
   lazy val vtTestVcfBiallelicExpectedNormalized =
     s"$testFolder/01_IN_altered_biallelic_bcftoolsnormalized.vcf"
 
+  lazy val vtTestVcfBiallelicExpectedSplitNormalized =
+    s"$testFolder/01_IN_altered_biallelic_vtdecompose_bcftoolsnormalized.vcf"
+
   lazy val vtTestVcfMultiAllelic =
-    s"$testFolder/01_IN_altered_multiallelic.vcf"
+    s"$testFolder/01_IN_altered_multiallelic_new.vcf"
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply core/src/test/scala/io/projectglow/transformers/normalizevariants/NormalizeVariantsTransformerSuite.scala]: new removed

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 16]
```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
+  }
+
+  /** calculates n choose r
+   *
+   * @param n: total number of objects
+   * @param r: number of selected objects
+   * @return n choose r
+   */
+  @VisibleForTesting
+  private[normalizevariants] def nChooseR(n: Int, r: Int): Int = {
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: No longer needed

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 9 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Forgot to delete.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 13 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: changed

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala (hunk 11 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: fixed.

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantSplitter.scala [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+              case f
+                  if structFieldsEqualExceptNullability(genotypeLikelihoodsField, f) |
+                  structFieldsEqualExceptNullability(phredLikelihoodsField, f) |
+                  structFieldsEqualExceptNullability(posteriorProbabilitiesField, f) =>
+                // update genotypes subfields that have colex order using the udf
+                df.withColumn(
+                  f.name,
+                  when(
+                    col(splitFromMultiAllelicField.name),
+                    expr( // TODO: Hard code the result of udf for numAlleles = 2, 3, 4 and ploidy = 2
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantSplitter.scala]: Old remnants...

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantSplitter.scala [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+import io.projectglow.vcf.InternalRowToVariantContextConverter
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.SQLUtils.structFieldsEqualExceptNullability
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+private[projectglow] object VariantSplitter extends GlowLogging {
+
+  /**
+   * Generates a new DataFrame by splitting the variants in the input DataFrame similar to what vt decompose does
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantSplitter.scala]: Done

File: core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala [hunk 19]
```diff
@@ -67,92 +66,66 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
...
-          schema,
-          validationStringency
-        )
+    // TODO: Implement normalization without using VariantContext
+    val dfAfterMaybeNormalize = if (doNormalize) {
+      val rddAfterNormalize = {
+        // The following if-statement is temporary to make the normalization work.
+        // After normalization is SQLified it will be dropped.
+        if (splitToBiallelic) {
+          dfAfterMaybeSplit.drop(infoFieldPrefix + oldMultiallelicFieldName)
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala]: Changed. It now works.

File: core/src/test/scala/io/projectglow/transformers/normalizevariants/VariantNormalizerSuite.scala [hunk 20]
```diff
@@ -107,4 +115,276 @@ class VariantNormalizerSuite extends GlowBaseTest with GlowLogging {
...
+    val dfExpected = spark
+      .read
+      .format(sourceName)
+      .load(vtTestVcfMultiAllelicExpectedSplit)
+      .orderBy("contigName", "start", "end")
+
+    assert(dfSplitInfo.count() == dfExpected.count())
+    assert(dfSplitInfo.count() == dfOriginal.count())
+
+// Check if Info Columns after splitting are the same as vt decompose
```
[the remark this answers is not in the document]
Comment (kianfar77) [reply core/src/test/scala/io/projectglow/transformers/normalizevariants/VariantNormalizerSuite.scala]: fixed

File: core/src/main/scala/io/projectglow/sql/optimizer/hlsOptimizerRules.scala [hunk 21]
```diff
@@ -52,3 +59,39 @@ object ResolveAggregateFunctionsRule extends Rule[LogicalPlan] {
       AggregateExpression(agg.asWrapped, Complete, isDistinct = false)
   }
 }
+
+/**
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/sql/optimizer/hlsOptimizerRules.scala]: Delete this if it doesn't help
